# Описание данных
# Таблица books
Содержит данные о книгах:
book_id — идентификатор книги;
author_id — идентификатор автора;
title — название книги;
num_pages — количество страниц;
publication_date — дата публикации книги;
publisher_id — идентификатор издателя.
# Таблица authors
Содержит данные об авторах:
author_id — идентификатор автора;
author — имя автора.
# Таблица publishers
Содержит данные об издательствах:
publisher_id — идентификатор издательства;
publisher — название издательства;
# Таблица ratings
Содержит данные о пользовательских оценках книг:
rating_id — идентификатор оценки;
book_id — идентификатор книги;
username — имя пользователя, оставившего оценку;
rating — оценка книги.
# Таблица reviews
Содержит данные о пользовательских обзорах:
review_id — идентификатор обзора;
book_id — идентификатор книги;
username — имя автора обзора;
text — текст обзора.



# Задания
1. Посчитайте, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитайте количество обзоров и среднюю оценку;
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.
# Как выполнить задание?
Опишите цели исследования;
Исследуйте таблицы — выведите первые строки;
Сделайте по одному SQL-запросу для решения каждого задания;
Выведите результаты каждого запроса в тетрадке;
Опишите выводы по каждой из решённых задач.

# 1. Посчитайте, сколько книг вышло после 1 января 2000 года;

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
              'pwd': 'Sdf4$2;d-d30pp', # пароль
              'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
              'port': 6432, # порт подключения
              'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Просмотрю все таблицы 

Таблица books

In [2]:
# создаём запрос
query = '''
SELECT *
FROM books
LIMIT 5;
'''

In [3]:
# получаем данные на основании запроса
data = pd.io.sql.read_sql(query, con = engine)

In [4]:
# выводим на экран
display(data)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

# Таблица содержит 
* book_id — идентификатор книги; 
* author_id — идентификатор автора; 
* title — название книги; 
* num_pages — количество страниц; 
* publication_date — дата публикации книги; 
* publisher_id — идентификатор издателя.

Таблица authors

In [5]:
# создаём запрос
query = '''
SELECT *
FROM authors
LIMIT 5;
'''

In [6]:
data = pd.io.sql.read_sql(query, con = engine)

In [7]:
display(data)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

# Таблица содержит
* author_id — идентификатор автора; 
* author — имя автора

Таблица publishers

In [8]:
query = '''
SELECT *
FROM publishers
LIMIT 5;
'''

In [9]:
data = pd.io.sql.read_sql(query, con = engine)

In [10]:
display(data)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

# Таблица содержит 
* publisher_id — идентификатор издательства; 
* publisher — название издательства;

Таблица  ratings

In [11]:
query = '''
SELECT *
FROM ratings
LIMIT 5;
'''

In [12]:
data = pd.io.sql.read_sql(query, con = engine)

In [13]:
display(data)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

# Таблица содержит
* rating_id — идентификатор оценки; 
* book_id — идентификатор книги; 
* username — имя пользователя, оставившего оценку; 
* rating — оценка книги

Таблица reviews

In [14]:
query = '''
SELECT *
FROM ratings
LIMIT 5;
'''

In [15]:
data = pd.io.sql.read_sql(query, con = engine)

In [16]:
display(data)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

# Таблица содержит
* review_id — идентификатор обзора; 
* book_id — идентификатор книги; 
* username — имя автора обзора; 
* text — текст обзора.

# Вывод
Базы все изучила, просмотрела.Все в продядке, как и заявленно в описании.


# 1. Посчитайте, сколько книг вышло после 1 января 2000 года;

In [17]:
query = '''
SELECT COUNT(book_id) AS количество_книг 
FROM books
WHERE publication_date::DATE > '2000-01-01';
'''

In [18]:
data = pd.io.sql.read_sql(query, con = engine)

In [19]:
display(data)

количество_книг
0              819

Вывод: 819 книг вышло после 1 января 2000 года

# 2. Для каждой книги посчитайте количество обзоров и среднюю оценку


In [20]:
query = '''
SELECT title AS кол_обзоров, средняя_оценка FROM books 
LEFT JOIN 
(SELECT book_id, COUNT(review_id) AS кол_обзоров FROM reviews GROUP BY book_id) AS count_rew
ON count_rew.book_id = books.book_id 
LEFT JOIN 
(SELECT book_id, AVG(rating) AS средняя_оценка FROM ratings GROUP BY book_id) AS avg_rating
ON books.book_id = avg_rating.book_id
'''

In [21]:
data = pd.io.sql.read_sql(query, con = engine)

In [22]:
display(data)

кол_обзоров  средняя_оценка
0            The Body in the Library (Miss Marple  #3)        4.500000
1                                            Galápagos        4.500000
2                             A Tree Grows in Brooklyn        4.250000
3    Undaunted Courage: The Pioneering First Missio...        4.000000
4                                          The Prophet        4.285714
..                                                 ...             ...
995                                Alice in Wonderland        4.230769
996          A Woman of Substance (Emma Harte Saga #1)        5.000000
997                                          Christine        3.428571
998  The Magicians' Guild (Black Magician Trilogy  #1)        3.500000
999                           The Plot Against America        3.000000

[1000 rows x 2 columns]

# Вывод
Для каждой книги посчитано количество обзоров и средний рейтинг

# 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [23]:
query = '''
SELECT
    publishers.publisher,
    COUNT(book_id)
FROM books INNER JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher
ORDER BY count DESC
LIMIT 1;
'''

In [24]:
data = pd.io.sql.read_sql(query, con = engine)

In [25]:
display(data)

publisher  count
0  Penguin Books     42

# Вывод

Самое продуктивное издательство - Penguin Books, оно выпустило 42 книги толще 50 стр.

# 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;


In [26]:
query = '''
SELECT
    author,
    AVG(subq.rating) AS avg_rating
FROM (
    SELECT
        author,
        title,
        AVG(rating) AS rating
    FROM (
        SELECT *
        FROM books
        WHERE book_id IN (
            SELECT book_id
            FROM ratings
            GROUP BY book_id
            HAVING COUNT(rating) >= 50
        ) 
    ) AS rated_books
    INNER JOIN authors ON rated_books.author_id = authors.author_id
    INNER JOIN ratings ON rated_books.book_id = ratings.book_id
    GROUP BY author, title
    ORDER BY author
) AS subq
GROUP BY author
ORDER BY avg_rating DESC
LIMIT 1
'''

In [27]:
data = pd.io.sql.read_sql(query, con = engine)

In [28]:
display(data)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

# Вывод
Дуэт писательницы Джоанны Роулинг и иллюстратора Мэри Гранпрэ имеет наивысший средний рейтинг написанных книг (4.28) среди книг, имеющих 50 и более оценок

# 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [29]:
# создаём запрос
query = '''
SELECT
    AVG(subq.review_count) AS average_review_count
FROM (
    SELECT COUNT(review_id) AS review_count
        FROM
            reviews
        WHERE username IN (
            SELECT
                username
            FROM ratings
            GROUP BY username
            HAVING COUNT(rating) > 50
            )
    GROUP BY username
    ) AS subq;
'''

In [30]:
data = pd.io.sql.read_sql(query, con = engine)

In [31]:
display(data)

average_review_count
0             24.333333

# Вывод
Среднее количество обзоров, написанных пользователями, которые поставили более 50 оценок, составляет 24.(3)

# Итог:
* Посчитала, сколько книг вышло после 1 января 2000 года - 819 книг;
* Для каждой книги посчитать количество обзоров и среднюю оценку - посчитала:
* Определить издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books;
* Определить автора с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré;
* Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24,3 
    